### PR TITLE
To add SequentialLR to PyTorch Core Schedulers

### LINKED ISSUE #4281: Update reference scripts to use the "Batteries Included" utils
## 🚀 Feature
As part of the "Batteries Included" initiative (#3911) we are adding a number of new utils that can be used to produce SOTA results. Once those utils are landed, we should update our reference scripts to use them. 

More specifically we need to:
- [ ] Update the reference scripts to use PyTorch's new [linear warmup](https://github.com/pytorch/pytorch/pull/60836) scheduler.
  - The reference scripts should be updated to let users define the `warmup_method` (if any), the `warmup_iters` and the `warmup_factor`. The warmup scheduler should be chained with other existing schedulers.
  - The Object Detection recipe needs to be BC compatible and replace our [custom linear warmup](https://github.com/pytorch/vision/blob/dc42f933f3343c76727dbfba6e4242f1bcb8e1a0/references/detection/engine.py#L20-L24) approach with the one from PyTorch.
  - Similarly the Video Classification recipe needs also to be BC compatible and replace our [customer scheduler](https://github.com/pytorch/vision/blob/master/references/video_classification/scheduler.py#L5).
  - All other recipes (Classification and Segmentation), should be updated to optionally use warm up. The addition should be BC compatible and turned off by default.
- [x] Update the Classification reference to use PyTorch's new label smoothing implementation.
  - We will support the case of label smoothing by using the upcoming `label_smoothing` argument of CrossEntropyLoss (https://github.com/pytorch/pytorch/pull/63122). 
  - A new `label_smoothing` argument should be introduced on the [script](https://github.com/pytorch/vision/blob/c2ab0c59f42babf9ad01aa616cd8a901daac86dd/references/classification/train.py#L165). To ensure BC, its default value should be 0.0 (no label smoothing).

### PR DESCRIPTION
**Pull request by iramazanli**
Partially resolves https://github.com/pytorch/vision/issues/4281

In this PR we are proposing a new scheduler --SequentialLR-- which enables list of different schedulers called in different periods of the training process.

The main motivation of this scheduler is recently gained popularity of warming up phase in the training time. It has been shown that having a small steps in initial stages of training can help convergence procedure get faster.

With the help of SequentialLR we mainly enable to call a small constant (or linearly increasing) learning rate followed by actual target learning rate scheduler.

```PyThon
scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=2)
scheduler2 = ExponentialLR(optimizer, gamma=0.9)
scheduler = SequentialLR(optimizer, schedulers=[scheduler1, scheduler2], milestones=[5])

for epoch in range(100):
    train(...)
    validate(...)
    scheduler.step()
```

which this code snippet will call `ConstantLR` in the first 5 epochs and will follow up with `ExponentialLR` in the following epochs.

This scheduler could be used to provide call of any group of schedulers next to each other. The main consideration we should make is every time we switch to a new scheduler we assume that new scheduler starts from the beginning- zeroth epoch.

We also add Chained Scheduler to `optim.rst` and `lr_scheduler.pyi` files here.

### CODE REVIEW

**Review by datumbox — APPROVED** [review]
LGTM. I left some nits on the comments for your consideration but it's up to you.

-----------

I only have a question about your example on the description of this PR. Here is what you have at the time of writing:

```python
scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=2)
scheduler2 = ExponentialLR(optimizer, gamma=0.9)
scheduler = SequentialLR(optimizer, schedulers=[scheduler1, scheduler2], milestones=[5])

for epoch in range(100):
    train(...)
    validate(...)
    scheduler.step()
```

The example has a funny warmup `total_iters=2` and `milestones=[5]` value mismatch (I believe it's just a typo) which raises an interesting question on how this scheduler works. My understanding is that for the first 2 epochs the warmup will use 10% of the LR, for 3 additional epochs it will be 100% of its value and then from 10 to 100 epoch the ExponentialLR scheduler will pick up. Is my understanding correct? If yes, I think that's the expected behaviour if someone misconfigures the scheduler.

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -582,6 +582,51 @@ def _get_closed_form_lr(self):
...
+
+    def __init__(self, optimizer, schedulers, milestones, last_epoch=-1, verbose=False):
+        for scheduler_idx in range(1, len(schedulers)):
+            if (schedulers[scheduler_idx].optimizer != schedulers[0].optimizer):
+                raise ValueError(
+                    "Sequential Schedulers expects all schedulers to belong to the same optimizer, but "
+                    "got schedulers at index {} and {} to be different".format(0, scheduler_idx)
+                )
+        self._schedulers = schedulers
+        self._milestones = milestones
```
Comment (fmassa) [inline torch/optim/lr_scheduler.py:618]: Should we add a check that `len(milestones) == len(schedulers) - 1`?

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!

Following @datumbox comment and for my own understanding, what happens if we have `milestones = 3` but use it with `ConstantLR` of size 5? Do we restart from the original LR when we switch schedulers or do we pick the smaller LR?

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (iramazanli) [reply torch/optim/lr_scheduler.py:618]: that's a really good idea ! thanks for pointing it out

File: test/test_optim.py [hunk 2]
```diff
@@ -1255,6 +1255,29 @@ def test_reduce_lr_on_plateau8(self):
                                       threshold=0.1, patience=5, cooldown=5)
         self._test_reduce_lr_on_plateau(scheduler, targets, metrics, epochs)
 
+    def test_sequentiallr1(self):
+        epochs = 10
```
[the remark this answers is not in the document]
Comment (iramazanli) [reply test/test_optim.py:1259]: oh thanks for pointing it out, i indeed changed this test several times, `epochs = 10` stayed from the previous version.

File: test/test_optim.py [hunk 3]
```diff
@@ -1255,6 +1255,29 @@ def test_reduce_lr_on_plateau8(self):
...
+                                       + [0.05 * 0.01 for x in range(4)]
+                                       + [0.05 * 0.001 for x in range(4)]]
+        milestones = [3]
+        schedulers[0] = ExponentialLR(self.opt, gamma=0.8)
+        schedulers[1] = StepLR(self.opt, gamma=0.1, step_size=4)
+        scheduler = SequentialLR(self.opt, schedulers=schedulers, milestones=milestones)
+        self._test(scheduler, targets, epochs)
+
+    def test_sequentiallr2(self):
+        epochs = 10
```
[the remark this answers is not in the document]
Comment (iramazanli) [reply test/test_optim.py]: test added with 3 schedulers :)

**Comment by iramazanli** [comment]
> LGTM. I left some nits on the comments for your consideration but it's up to you.
> 
> I only have a question about your example on the description of this PR. Here is what you have at the time of writing:
> 
> ```python
> scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=2)
> scheduler2 = ExponentialLR(optimizer, gamma=0.9)
> scheduler = SequentialLR(optimizer, schedulers=[scheduler1, scheduler2], milestones=[5])
> 
> for epoch in range(100):
>     train(...)
>     validate(...)
>     scheduler.step()
> ```
> 
> The example has a funny warmup `total_iters=2` and `milestones=[5]` value mismatch (I believe it's just a typo) which raises an interesting question on how this scheduler works. My understanding is that for the first 2 epochs the warmup will use 10% of the LR, for 3 additional epochs it will be 100% of its value and then from 10 to 100 epoch the ExponentialLR scheduler will pick up. Is my understanding correct? If yes, I think that's the expected behaviour if someone misconfigures the scheduler.

Your explanation for the expected behavior is correct, and it's not necessary the typo. because theoretically it could be configured that way without any error.

Given that there can be any kind of schedulers added to SequentialLR i preferred to let all the schedulers behave in the default mode. Please let me know if you have a feedback about it.

**Comment by iramazanli** [comment]
> Thanks for the PR!
> 
> Following @datumbox comment and for my own understanding, what happens if we have `milestones = 3` but use it with `ConstantLR` of size 5? Do we restart from the original LR when we switch schedulers or do we pick the smaller LR?

The way it behaves is as 3 epoch it behave like ConstantLR, then when we move to new scheduler it just behaves like it started from scratch in the 4-th epoch.
